control-plane/connect-inject/peering_dialer_controller_test.go

```diff
@@ -18,6 +23,462 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client/fake"
 )
 
+// TestReconcileCreateUpdatePeeringDialer creates a peering dialer.
+func TestReconcileCreateUpdatePeeringDialer(t *testing.T) {
+	t.Parallel()
+	nodeName := "test-node"
+	node2Name := "test-node2"
+	cases := map[string]struct {
+		peeringName            string
+		k8sObjects             func() []runtime.Object
+		expectedConsulPeerings *api.Peering
+		peeringSecret          func(token string) *corev1.Secret
+		expErr                 string
+		expectedStatus         *v1alpha1.PeeringDialerStatus
+		expectDeletedK8sSecret *types.NamespacedName
+		peeringExists          bool
+	}{
+		"Errors when Secret is not set on the spec": {
+			k8sObjects: func() []runtime.Object {
+				dialer := &v1alpha1.PeeringDialer{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "peering",
+						Namespace: "default",
+					},
+					Spec: v1alpha1.PeeringDialerSpec{
+						Peer: &v1alpha1.Peer{
+							Secret: nil,
+						},
+					},
+				}
+				return []runtime.Object{dialer}
+			},
+			expErr:        "PeeringDialer spec.peer.secret was not set",
+			peeringSecret: func(_ string) *corev1.Secret { return nil },
+		},
+		"Errors when Secret set on the spec does not exist in the cluster": {
+			k8sObjects: func() []runtime.Object {
+				dialer := &v1alpha1.PeeringDialer{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "peering",
+						Namespace: "default",
+					},
+					Spec: v1alpha1.PeeringDialerSpec{
+						Peer: &v1alpha1.Peer{
+							Secret: &v1alpha1.Secret{
+								Name:    "dialer",
+								Key:     "token",
+								Backend: "kubernetes",
+							},
+						},
+					},
+				}
+				return []runtime.Object{dialer}
+			},
+			expErr:        "PeeringDialer spec.peer.secret does not exist",
+			peeringSecret: func(_ string) *corev1.Secret { return nil },
+		},
+		"Initiates peering when status secret is nil": {
+			peeringName: "peering",
+			k8sObjects: func() []runtime.Object {
+				dialer := &v1alpha1.PeeringDialer{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "peering",
+						Namespace: "default",
+					},
+					Spec: v1alpha1.PeeringDialerSpec{
+						Peer: &v1alpha1.Peer{
+							Secret: &v1alpha1.Secret{
+								Name:    "dialer-token",
+								Key:     "token",
+								Backend: "kubernetes",
+							},
+						},
+					},
+				}
+				return []runtime.Object{dialer}
+			},
+			expectedConsulPeerings: &api.Peering{
+				Name:  "peering",
+				State: api.PeeringStateActive,
+			},
+			peeringSecret: func(token string) *corev1.Secret {
+				return createSecret("dialer-token", "default", "token", token)
```

Comment:
Could it make sense to set the resource version on the secret here, and assert on that resource version in the expectedStatus? We might be able to do it here since we control the peering secret.